### ISSUE TITLE
Add a continuous fuzz testing suite

### ISSUE THREAD

**Issue by silvergasp** [biopython/biopython#4223]
I've written a couple of fuzz tests and found a few (non-critical) bugs in some of the parsers. Before continuing any further is there any interest in accepting a PR with fuzz tests?

If you aren't familiar with fuzz testing, here is a bit of a run down ([from Wikipedia](https://en.wikipedia.org/wiki/Fuzzing));

> In [programming](https://en.wikipedia.org/wiki/Computer_programming) and [software development](https://en.wikipedia.org/wiki/Software_development), fuzzing or fuzz testing is an automated [software testing](https://en.wikipedia.org/wiki/Software_testing) technique that involves providing invalid, unexpected, or [random data](https://en.wikipedia.org/wiki/Random_data) as inputs to a [computer program](https://en.wikipedia.org/wiki/Computer_program). The program is then monitored for exceptions such as [crashes](https://en.wikipedia.org/wiki/Crash_(computing)), failing built-in code [assertions](https://en.wikipedia.org/wiki/Assertion_(software_development)), or potential [memory leaks](https://en.wikipedia.org/wiki/Memory_leak).

If this sounds like it'd be useful, as a follow-up step I'd like to suggest adding biopython to google/oss-fuzz. OSS-fuzz is Google's free open-source continuous fuzzing service. If biopython is integrated into oss-fuzz, the fuzz tests under biopython will be built and then run once a day, to search for bugs and vulnerabilities in biopython. This service can be integrated with the CI for biopython, so that the fuzz tests are run for 10min or so for every pull request, preventing buggy code from being merged.

**Comment by peterjc** [comment]
This seems worth trying, although I think it would be most valuable for parsers for the well defined file formats. On the more ambiguous text formats I fear the fuzz testing could flag an unmanageable number of theoretical concerns. My hunch could be wrong though.

Exploring fuzz testing outside of CI would seem a sensible first step - there are likely more low hanging fruit (on top of the issues you alluded to) which you'd need to fix before using this within CI would be helpful in preventing new issues.

**Comment by silvergasp** [comment]
> This seems worth trying, although I think it would be most valuable for parsers for the well defined file formats. On the more ambiguous text formats I fear the fuzz testing could flag an unmanageable number of theoretical concerns. My hunch could be wrong though.

Yeah, I'd agree with that, there seems to be a fair bit of work to be done to make the PDB parser robust against arbitrary inputs. I think you are right that the number of bugs the fuzzer seems to find would be unmanageable if plugged into the CI. I'm not super familiar with the underlying structure of each format, so If you have an idea of which formats are "well defined" I'd be happy to write a fuzzer for those as a starting point.

> Exploring fuzz testing outside of CI would seem a sensible first step - there are likely more low hanging fruit (on top of the issues you alluded to) which you'd need to fix before using this within CI would be helpful in preventing new issues.

I've also been experimenting with adding a fuzzing harness for the Python functions that call C functions. In particular, fuzzing the `PDB.Superimposer`. At the moment the superimposer API seems to be fairly robust to arbitrary inputs. i.e. the fuzzer doesn't crash within the first hour or so of running, compared to the parser fuzzer which crashes in seconds. The reason I chose this API is that if there is a bug in the C code it's much more likely to be a security issue than a bug in the python code. I'll open a PR for that fuzz harness once I'm done cleaning up a little. I think that the superimposer API is at a point where it would benefit from running in the CI, as there aren't any easily detectable bugs :)

**Comment by peterjc** [comment]
From the security point of view, fuzzing the C code in general might be quite a sensible first target (see modules in ``setup.py`` for a list). e.g. the pairwise alignment code which is relatively young. There is also a C code version of the Nexus parser (which at this point is pretty well field tested).

**Comment by silvergasp** [comment]
Ok so I've added a fuzz harness https://github.com/biopython/biopython/pull/4225 PR for PDB.Superimposer which wraps the C api. I think that the Superimposer is quite robust, and could benefit from having the fuzz tests in CI or integrated into the OSS-fuzz continuous integration suite.
